controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +108,293 @@ func NewReconciler(
+}
+
+// reconcileNewPublishRequests compares a set of exist requests and a set of expected requests and
+// creates new requests if they are missing in the exist requests set.
+// It returns nil ONLY when no new publish requests were created, in other words, the set of exist requests
+// and the set of expected requests are the same.
+func (r *Reconciler) reconcileNewPublishRequests(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext,
+	existReqsSet, desiredReqsSet sets.Set[string]) error {
+
+	newReqsSet := desiredReqsSet.Difference(existReqsSet)
+	if newReqsSet.Len() == 0 {
+		return nil
+	}
+
+	vmGroupPubReqGvk, err := apiutil.GVKForObject(ctx.VMGroupPublishRequest, r.Scheme())
+	if err != nil {
+		return err
+	}
+
+	for sourceVM := range newReqsSet {
+		// TTLSecondsAfterFinished is intentionally unset to avoid vm publish request to be
+		// deleted after completion.
+		if err := r.Create(ctx, &vmopv1.VirtualMachinePublishRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    ctx.VMGroupPublishRequest.Namespace,
+				GenerateName: ctx.VMGroupPublishRequest.Name + "-",
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(ctx.VMGroupPublishRequest, vmGroupPubReqGvk),
+				},
+			},
+			Spec: vmopv1.VirtualMachinePublishRequestSpec{
+				Source: vmopv1.VirtualMachinePublishRequestSource{
+					Name:       sourceVM,
+					Kind:       reflect.TypeOf(vmopv1.VirtualMachine{}).Name(),
+					APIVersion: vmopv1.GroupVersion.String(),
+				},
+				Target: vmopv1.VirtualMachinePublishRequestTarget{
+					Location: vmopv1.VirtualMachinePublishRequestTargetLocation{
+						Name:       ctx.VMGroupPublishRequest.Spec.Target,
+						Kind:       reflect.TypeOf(imgregv1a1.ContentLibrary{}).Name(),
+						APIVersion: imgregv1a1.GroupVersion.String(),
+					},
+				},
+			},
+		}); err != nil {
+			return err
+		}
+	}
+
+	return pkgerr.RequeueError{}
+}
+
+// reconcileStatusCompletedCondition sets vm group publish request's completed condition and completion time
+// once all desired vm publish requests are completed.
+func (r *Reconciler) reconcileStatusCompletedCondition(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext,
+	pendingReqsSet, completedReqsSet, desiredReqsSet sets.Set[string]) error {
+
+	if completedReqsSet.Equal(desiredReqsSet) {
+		conditions.MarkTrue(ctx.VMGroupPublishRequest, vmopv1.VirtualMachineGroupPublishRequestConditionComplete)
+		ctx.VMGroupPublishRequest.Status.CompletionTime = metav1.Now()
+		// requeue immediately for processing TTL
+		return pkgerr.RequeueError{}
+	}
+
+	conditions.MarkFalse(
+		ctx.VMGroupPublishRequest,
+		vmopv1.VirtualMachinePublishRequestConditionComplete,
+		vmopv1.VirtualMachineGroupPublishRequestConditionReasonPending,
+		"waiting %d more vm publish requests to be completed",
+		pendingReqsSet.Len())
+
+	// requeue for default sync period while wait for more requests to be completed
+	return pkgerr.RequeueError{After: pkgmgr.DefaultSyncPeriod}
```

Comment:
This is the full resync period so it doesn't make sense to requeue after that period. What this controller should be doing is using EnqueueRequestForOwner - see https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/handler#example-EnqueueRequestForOwner and existing usages in our code.

You should be able to remove the `forceReconcile()` func in the test code after this change

Reply:
TIL. will do, thanks!